### PR TITLE
Inverted index hotbackup

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR: https://github.com/arangodb/enterprise/pull/1052
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: client-tools/Restore/RestoreFeature.cpp [hunk 1]
```diff
@@ -821,20 +821,31 @@ arangodb::Result processInputDirectory(
...
+          if (auto r = ::restoreView(httpClient, options, slice); !r.ok()) {
+            return r;
+          }
         }
       }
+      return Result{};
+    };
+
+    // Step 5: create arangosearch views
+    auto r = createViews("arangosearch");
```
Comment (gnusi) [inline client-tools/Restore/RestoreFeature.cpp:846]: `StaticStrings::`?

File: client-tools/Restore/RestoreFeature.cpp [hunk 2]
```diff
@@ -884,6 +895,12 @@ arangodb::Result processInputDirectory(
     jobQueue.waitForIdle();
     jobs.clear();
 
+    // Step 7: create search-alias views
+    r = createViews("search-alias");
```
Comment (gnusi) [inline client-tools/Restore/RestoreFeature.cpp:899]: `StaticStrings::`?

File: arangod/IResearch/Search.cpp [hunk 3]
```diff
@@ -642,7 +632,11 @@ Result Search::updateProperties(CollectionNameResolver& resolver,
       frozen::make_unordered_set<frozen::string>({"", "add", "del"});
   for (; it.valid(); ++it) {
     auto value = *it;
-    auto collection = getCollection(resolver, value.get("collection"));
+    auto collectionSlice = value.get("collection");
+    if (!collectionSlice.isString()) {
+      return {TRI_ERROR_BAD_PARAMETER, "'index' should be string"};
```
Comment (gnusi) [inline arangod/IResearch/Search.cpp:637]: `a string`

File: arangod/IResearch/Search.cpp [hunk 4]
```diff
@@ -671,7 +665,11 @@ Result Search::updateProperties(CollectionNameResolver& resolver,
                 "Cannot find collection for index to delete"};
       }
     }
-    auto index = getIndex(*collection, value.get("index"));
+    auto indexSlice = value.get("index");
+    if (!indexSlice.isString()) {
+      return {TRI_ERROR_BAD_PARAMETER, "'index' should be string"};
```
Comment (gnusi) [inline arangod/IResearch/Search.cpp:670]: `a string`

File: arangod/IResearch/Search.cpp [hunk 5]
```diff
@@ -573,23 +556,30 @@ Result Search::appendVPackImpl(velocypack::Builder& build, Serialization ctx,
...
           } else {
             build.add(velocypack::Value{velocypack::ValueType::Object});
-            build.add("collection", velocypack::Value{cid.id()});
-            // TODO Maybe guid or planId?
-            build.add("index", velocypack::Value{index->id().id()});
+            if (ServerState::instance()->isCoordinator()) {
+              build.add("collection", velocypack::Value{collection->name()});
+            } else {
+              TRI_ASSERT(ServerState::instance()->isSingleServer());
+              build.add("collection", velocypack::Value{collection->guid()});
```
Comment (gnusi) [inline arangod/IResearch/Search.cpp:579]: comment?

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 6]
```diff
@@ -1923,6 +1923,11 @@ Result RestReplicationHandler::processRestoreIndexes(
         if (e.code() == TRI_ERROR_NOT_IMPLEMENTED) {
           continue;
         }
+        if (auto const& message = e.message();
```
Comment (Dronplane) [inline arangod/RestHandler/RestReplicationHandler.cpp:1926]: Could we at least check exception code here? Like `TRI_ERROR_ARANGO_DUPLICATE_NAME` or smth

**Review by Dronplane — APPROVED** [review]
LGTM

File: tests/js/server/dump/dump-test.inc [hunk 7]
```diff
@@ -835,6 +835,51 @@
...
+        var startTime = new Date();
+        var res;
+        while (new Date() - startTime < 60000) {
+          res = db._query("FOR doc IN " + search.name() + " SEARCH doc.value >= 0 OPTIONS { waitForSync: true } RETURN doc").toArray();
+          if (res.length === 5000) {
+            break;
+          }
+          console.log("Waiting for inverted index to be built...");
+          internal.wait(1);
+        }
```
Comment (gnusi) [inline tests/js/server/dump/dump-test.inc:870]: we don't need this anymore in presence of waitForSync
```
db._query("FOR doc IN " + search.name() + " SEARCH doc.value >= 0 OPTIONS { waitForSync: true } RETURN doc")
```

**Review by gnusi — APPROVED** [review]

File: arangod/IResearch/IResearchRocksDBInvertedIndex.cpp [hunk 8]
```diff
@@ -95,6 +95,15 @@ std::shared_ptr<Index> IResearchRocksDBInvertedIndexFactory::instantiate(
...
   auto index = std::make_shared<IResearchRocksDBInvertedIndex>(
       id, collection, objectId, indexName);
+  auto& cf = _server.getFeature<ClusterFeature>();
+  if (cf.isEnabled()) {
+    std::string collectionName;
+    clusterCollectionName(
+        collection, clusterWideIndex ? nullptr : &cf.clusterInfo(),
+        index->Index::id().id(), false /*TODO meta.willIndexIdAttribute()*/,
+        collectionName);
+    index->setCollectionName(std::move(collectionName));
```
Comment (Dronplane) [inline arangod/IResearch/IResearchRocksDBInvertedIndex.cpp:105]: Do we really need to request Agency each time? Maybe check if it is already set in meta and use that value?

Comment (MBkkt) [reply arangod/IResearch/IResearchRocksDBInvertedIndex.cpp]: Done

File: tests/js/server/dump/dump-test.inc (hunk 7 above)
Comment (MBkkt) [reply tests/js/server/dump/dump-test.inc:870]: Old code. I will change it if build failed

**Pull request closed without merging**